contracts/drafts/meta-tx/GSNContext.sol

```diff
@@ -0,0 +1,71 @@
+pragma solidity ^0.5.0;
+
+import "./Context.sol";
+
+/*
+ * @dev Enables GSN support by recognizing calls from RelayHub and extracting
+ * the actual sender and call data from the received values.
+ */
+contract GSNContext is Context {
+    address private _relayHub;
+
+    constructor(address relayHub) public {
+        _setRelayHub(relayHub);
+    }
+
+    function _setRelayHub(address relayHub) internal {
+        _relayHub = relayHub;
+    }
+
+    // Overrides for Context's functions: when called from RelayHub, sender and
+    // data require some pre-processing: the actual sender is stored at the end
+    // of the call data, which in turns means it needs to be removed from it
+    // when handling said data.
+
+    function _msgSender() internal view returns (address) {
+        if (msg.sender != _relayHub) {
+            return msg.sender;
+        } else {
+            return _getRelayedCallSender();
+        }
+    }
+
+    function _msgData() internal view returns (bytes memory) {
+        if (msg.sender != _relayHub) {
+            return msg.data;
+        } else {
+            return _getRelayedCallData();
+        }
+    }
+
+    function _getRelayedCallSender() private pure returns (address result) {
+        // We need to read 20 bytes (an address) located at array index msg.data.length - 20. In memory, the array
+        // is prefixed with a 32-byte length value, so we first add 32 to get the memory read index. However, doing
+        // so would leave the address in the upper 20 bytes of the 32-byte word, which is inconvenient and would
+        // require bit shifting. We therefore subtract 12 from the read index so the address lands on the lower 20
+        // bytes. This can always be done due to the 32-byte prefix.
+
+        // The final memory read index is msg.data.length - 20 + 32 - 12 = msg.data.length
+
+        // These fields are not accessible from assembly
+        bytes memory array = msg.data;
+        uint256 index = msg.data.length;
+
+        assembly {
+            // Load the 32 bytes word from memory with the address on the lower 20 bytes, and mask those.
+            result := and(mload(add(array, index)), 0xffffffffffffffffffffffffffffffffffffffff)
+        }
+        return result;
+    }
+
+    function _getRelayedCallData() private pure returns (bytes memory) {
```

Comment:
Should we cache the return value somehow?

Reply:
Wouldn't that require creating an extra copy, and copying back from it on each call? If we manage to get this function to a simple copy operation, then I wouldn't bother.

Reply:
Oh, because we cannot return the same array because it could've been modified? Good point, I hadn't thought of that. I was thinking we could return the same array always.

Reply:
Mutability was my concern, yes. I guess this ultimately boils down to how complex stripping the last 20 bytes is, so I'd defer this until we've settled on an implementation.